File: frag-nav/src/main/java/com/ncapdevi/fragnav/tabhistory/UnlimitedTabHistoryController.java
package com.ncapdevi.fragnav.tabhistory;

import android.support.annotation.NonNull;

import com.ncapdevi.fragnav.FragNavPopController;
import com.ncapdevi.fragnav.FragNavSwitchController;

import java.util.ArrayList;
import java.util.Stack;

public class UnlimitedTabHistoryController extends CollectionFragNavTabHistoryController {
    private Stack<Integer> tabHistory = new Stack<>();

    public UnlimitedTabHistoryController(FragNavPopController fragNavPopController,
                                         FragNavSwitchController fragNavSwitchController) {
        super(fragNavPopController, fragNavSwitchController);
    }

    @Override
    int getCollectionSize() {
        return tabHistory.size();
    }

    @Override
    int getAndRemoveIndex() {
        tabHistory.pop();
        return tabHistory.pop();
    }

    @Override
    public void switchTab(int index) {
        tabHistory.push(index);
    }

    @NonNull
    @Override
    ArrayList<Integer> getHistory() {
        return new ArrayList<>(tabHistory);
    }

    @Override
    void setHistory(@NonNull ArrayList<Integer> history) {
        tabHistory.clear();
        tabHistory.addAll(history);
    }
}

File: frag-nav/src/main/java/com/ncapdevi/fragnav/tabhistory/UnlimitedTabHistoryController.kt
package com.ncapdevi.fragnav.tabhistory

import com.ncapdevi.fragnav.FragNavPopController
import com.ncapdevi.fragnav.FragNavSwitchController
import java.util.*

class UnlimitedTabHistoryController(fragNavPopController: FragNavPopController,
                                    fragNavSwitchController: FragNavSwitchController) : CollectionFragNavTabHistoryController(fragNavPopController, fragNavSwitchController) {
    private val tabHistory = Stack<Int>()

    override val collectionSize: Int
        get() = tabHistory.size

    override val andRemoveIndex: Int
        get() {
            tabHistory.pop()
            return tabHistory.pop()
        }

    override var history: ArrayList<Int>
        get() = ArrayList(tabHistory)
        set(history) {
            tabHistory.clear()
            tabHistory.addAll(history)
        }

    override fun switchTab(index: Int) {
        tabHistory.push(index)
    }
}
